File: 23-Observer/src/go/weather_station.go
package main

import "fmt"

// WeatherStation is the Concrete Subject — it holds the temperature and pushes
// updates to every registered observer. It knows only the Observer interface.
type WeatherStation struct {
	observers   []Observer
	temperature float64
}

func (w *WeatherStation) SetTemperature(temperature float64) {
	w.temperature = temperature
	fmt.Printf("WeatherStation: temperature -> %.1f°C\n", temperature)
	w.NotifyObservers()
}

func (w *WeatherStation) RegisterObserver(observer Observer) {
	w.observers = append(w.observers, observer)
}

func (w *WeatherStation) RemoveObserver(observer Observer) {
	for i, o := range w.observers {
		if o == observer {
			w.observers = append(w.observers[:i], w.observers[i+1:]...)
			return
		}
	}
}

func (w *WeatherStation) NotifyObservers() {
	// Iterate a copy so an observer may unsubscribe during notification.
	snapshot := make([]Observer, len(w.observers))
	copy(snapshot, w.observers)
	for _, observer := range snapshot {
		observer.Update(w.temperature)
	}
}

File: 23-Observer/src/rust/weather_station.rs
use std::rc::Rc;

use crate::observer::Observer;

/// Concrete Subject — holds the temperature and pushes updates to every
/// registered observer. Observers are shared trait objects (`Rc<dyn Observer>`).
pub struct WeatherStation {
    observers: Vec<Rc<dyn Observer>>,
    temperature: f64,
}

impl WeatherStation {
    pub fn new() -> Self {
        WeatherStation { observers: Vec::new(), temperature: 0.0 }
    }

    pub fn set_temperature(&mut self, temperature: f64) {
        self.temperature = temperature;
        println!("WeatherStation: temperature -> {:.1}°C", temperature);
        self.notify_observers();
    }

    pub fn register_observer(&mut self, observer: Rc<dyn Observer>) {
        self.observers.push(observer);
    }

    pub fn remove_observer(&mut self, observer: &Rc<dyn Observer>) {
        // Compare by pointer identity to unsubscribe the exact observer.
        self.observers.retain(|o| !Rc::ptr_eq(o, observer));
    }

    pub fn notify_observers(&self) {
        // Clone the handles (cheap Rc bumps) so an observer could unsubscribe
        // during notification without corrupting the loop.
        let snapshot: Vec<Rc<dyn Observer>> = self.observers.clone();
        for observer in snapshot {
            observer.update(self.temperature);
        }
    }
}
